File: MergeKLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class MergeKLists {
    public ListNode mergeKLists(ListNode[] lists) {
        return merge(lists, 0, lists.length - 1);
    }
    private ListNode merge(ListNode[] lists, int start, int end){
        if(start == end) return lists[start];
        if(start > end) return null;
        int mid = start + (end - start)/2;
        return mergeTwoLists(merge(lists, start, mid), merge(lists, mid+1, end));
    }

    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        else if (l2 == null) return l1;

        ListNode protect = new ListNode(0, null);
        ListNode head = protect;

        while(l1 != null && l2 != null){
            if(l1.val > l2.val){
                protect.next = l2;
                protect = protect.next;
                l2 = l2.next;
            }else{
                protect.next = l1;
                protect = protect.next;
                l1 = l1.next;
            }
        }

        if (l2 != null) protect.next = l2;
        else if (l1 != null) protect.next = l1;
        return head.next;  
    }

}

File: MergeKLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
 #include<queue>

 struct Node{
    int key;
    ListNode* listNode;
    Node(int key, ListNode* listNode): key(key), listNode(listNode){};
 };

 bool operator < (const Node& a, const Node& b){
    //  a.key < b.key 大根堆
    //  a.key > b.key 小根堆
    return a.key > b.key;
 }

class MergeKLists {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<Node> q;  // 默认为大根堆
        for(ListNode* listNode: lists){
            if(listNode != nullptr) q.push(Node(listNode->val, listNode));
        }

        // 保护节点
        ListNode head;
        ListNode* last = &head;

        while(!q.empty()){
            Node node = q.top();
            q.pop();
            last->next = node.listNode;
            last = last->next;

            ListNode* p = node.listNode->next;
            if(p != nullptr) q.push(Node(p->val, p));
        }
        return head.next;
    }
};